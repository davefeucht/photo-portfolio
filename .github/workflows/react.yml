# This workflow will do a clean install of node dependencies, build the source code and run tests across different versions of node
# For more information see: https://help.github.com/actions/language-and-framework-guides/using-nodejs-with-github-actions

name: React CI/CD

on:
  push:
    branches: [ master ]
  pull_request:
    branches: [ master ]

jobs:
  build:

    runs-on: ubuntu-latest

    strategy:
      matrix:
        node-version: [18.x]

    steps:
    - name: Checkout code
      uses: actions/checkout@v2

    - name: Install Node.js ${{ matrix.node-version }}
      uses: actions/setup-node@v1
      with:
        node-version: ${{ matrix.node-version }}

    - name: Enable corepack
      run: corepack enable

    - name: Install dependencies
      run: yarn install

    - name: Build project
      run: npm run build --if-present

    - name: Run tests
      run: npm test
