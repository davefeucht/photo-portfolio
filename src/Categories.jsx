import React from 'react';
import ReactDOM from 'react-dom';
import axios from 'axios';
import Category from './Category.jsx';

class Categories extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      categories: [],
      showAllCategories: true,
      singleCategoryToShow: {id: 0, name: ""},
      errorMsg: ""
    }
  }

  _getCategories() {
    axios.get('http://' + this.props.site + '/wp-json/wp/v2/categories?exclude=1')
    .then(res => {
      const categories = res.data;
      this.setState({ categories });
    }, error => {
      const errorMsg = 'Did not work: ' + (error.response ? error.response : error);
      this.setState({ errorMsg });
    });
  }

  _showSpecificCategory(categoryId, categoryName) {
    this.setState({showAllCategories: !this.state.showAllCategories}); 
    this.setState({singleCategoryToShow: {id: categoryId, name: categoryName}}); 
  }

  componentWillMount() {
    this._getCategories();
  }
  
  render() {
    let categoryList = [];
    if(this.state.showAllCategories) {
      categoryList = this.state.categories.map(category =>
        { return ( <Category key={category.id.toString()} id={category.id} name={category.name} site={this.props.site} clickCategory={this._showSpecificCategory.bind(this)} showAllPosts={false} /> ); }
      );
    }
    else {
      categoryList = <Category key={this.state.singleCategoryToShow.id.toString()} id={this.state.singleCategoryToShow.id} name={this.state.singleCategoryToShow.name} site={this.props.site} clickCategory = {this._showSpecificCategory.bind(this)} showAllPosts={true} />;
    }
    return (
      <div className="category-list">
        {categoryList} 
        <div>{this.state.errorMsg}</div>
      </div>
    );
  }
}

export default Categories;
