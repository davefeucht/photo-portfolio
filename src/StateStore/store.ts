import { configure, makeAutoObservable } from 'mobx';
import { ImageData, Image, Store, Category, Post, Page } from '../utils/types';

configure({ enforceActions: 'observed' });

class stateStore implements Store {
    menuState = 'closed'
    applicationRoot: HTMLElement = null;
    screenInfo: {
        width: number,
        height: number
    }
    siteInfo: {
        siteName: string,
        siteUrl: string
    }
    visiblePost: {
        postId: number,
        postTitle: string,
        tags: number[],
        tagNames: string[],
        fullImageUrl: string,
        width: number,
        height: number
    };
    categoryList: Category[] = [];
    currentCategoryPosts: Post[] = [];
    currentCategoryData: Category = {
        id: 0,
        count: 0,
        description: '',
        link: '',
        name: '',
        slug: '',
        taxonomy: '',
        parent: 0
    };
    pages: Page[] = [];
    currentPageData: Page = {
        id: 0,
        link: '',
        modified: '',
        slug: '',
        type: '',
        title: {
            rendered: ''
        },
        content: {
            rendered: '',
            protected: false
        },
        author: 0,
        featured_media: 0
    };
    maxItemsPerPage = 10;

    constructor() {
        makeAutoObservable(this);
    }

    setMenuState = (state: string) => {
        this.menuState = state;
    }

    toggleMenuState = () => {
        this.menuState = this.menuState === 'closed' ? 'open' : 'closed';
    }

    setApplicationRoot = (element: HTMLElement) => {
        this.applicationRoot = element;
    }

    setSiteName = (name: string) => {
        this.siteInfo.siteName = name;
    }

    setVisiblePost = (postId: number, postTitle: string) => {
        this.visiblePost.postId = postId;
        this.visiblePost.postTitle = postTitle;
    }

    setVisiblePostImage = (fullImageUrl: string) => {
        this.visiblePost.fullImageUrl = fullImageUrl;
    }

    setVisiblePostTags = (tagNames: string[]) => {
        tagNames.forEach(tagName => {
            this.visiblePost.tagNames.push(tagName);
        });
    }

    clearVisiblePostTagNames = () => {
        this.visiblePost.tagNames.splice(0, this.visiblePost.tagNames.length);
    }

    setCategoryList = (categories: Category[]) => {
        this.categoryList.length = 0;
        categories.forEach((category, index) => {
            this.categoryList[index] = category;
        })
    }

    setCategoryPosts = (posts: Post[]) => {
        this.currentCategoryPosts.length = 0;
        posts.forEach((post, index) => {
            this.currentCategoryPosts[index] = post;
        })
    }

    setCategoryData = (categoryData: Category) => {
        Object.keys(categoryData).forEach((key: keyof Category) => {
            this.currentCategoryData[key] = categoryData[key];
        });
    }

    setThumbnailImageUrl = (imageData: ImageData) => {
        this.currentCategoryPosts[imageData.post_index].thumbnail_image = imageData.image_url;
    }

    setCurrentPost = (postData: Post) => {
        Object.keys(postData).forEach((key: keyof Image) => {
            this.visiblePost[key] = postData[key];
        })
    }

    setPages = (pages: Page[]) => {
        pages.forEach(page => {
            this.pages.push(page);
        });
    }

    setPageData = (pageData: Page) => {
        Object.assign(this.currentPageData, pageData);
    }
}

export default stateStore;
