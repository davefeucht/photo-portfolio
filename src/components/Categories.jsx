/*****************
* Categories component displays a list of existing photo categories 
* and makes them clickable to display just a single category.
*****************/

import React from "react";
import axios from "axios";
import CategoryList from "./styledComponents/CategoryList.jsx";
import CategoryThumbnail from "./CategoryThumbnail.jsx";
import PropTypes from "prop-types";

export default class Categories extends React.Component {

  static propTypes = {
    site: PropTypes.string,
    showSingleCategory: PropTypes.func
  };

  /* Add Default Props */

  //Set up state variables for list of categories, whether to show all categories, and if not, which
  //single category to show.
  //NOTE: Take list of categories out of state and store in a member variable.
  state = {
    categories: [],
    errorMsg: ""
  };

  _isMounted = false;

  //Method to get all existing categories
  _getCategories() {
    const getCategoriesRequest = `https://${this.props.site}/wp-json/wp/v2/categories?exclude=175`;
    if(this._isMounted) {
      axios.get(getCategoriesRequest)
        .then(res => {
          const categories = res.data;
          this.setState({ categories });
        })
        .catch(error => {
          const errorMsg = "Did not work: " + (error.response ? error.response : error);
          this.setState({ errorMsg });
        });
    }
  }

  _showSpecificCategory(categoryId, categoryName) {
    this.props.showSingleCategory(categoryId, categoryName);
  }

  //Just before the component is added to the DOM, get the list of categories.
  componentWillMount() {
    this._isMounted = true;
    this._getCategories();
  }

  componentWillUnmount() {
    this._isMounted = false;
  }
  
  render() {
    let categoryList = [];

    categoryList = this.state.categories.map(category =>
    { 
      //Display the Category component, and pass along the showSpecificCategory function as a prop, so that we can call it from the Category component
      return ( <CategoryThumbnail key={category.id.toString()} id={category.id} name={category.name} site={this.props.site} clickCategory={this._showSpecificCategory.bind(this)} /> ); }
    );

    return (
      <CategoryList>
        {categoryList} 
      </CategoryList>
    );
  }
}