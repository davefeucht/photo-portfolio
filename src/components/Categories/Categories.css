.category-list {
  min-height: 100%;
  display: grid;
  grid-template-columns: repeat(var(--number-of-columns), calc(100% / var(--number-of-columns)));
  grid-template-rows: repeat(var(--number-of-rows), 1fr);
}