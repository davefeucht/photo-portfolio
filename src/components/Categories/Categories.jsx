/*****************
* Categories component displays a list of existing photo categories 
* and makes them clickable to display just a single category.
*****************/

import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { reaction } from 'mobx';
import { observer } from 'mobx-react';
import CategoryThumbnail from '../CategoryThumbnail/CategoryThumbnail.jsx';
import './Categories.css';

const Categories = ({ stateStore, api }) => {
  const setRows = () => {
    const numberOfColumns = document.body.style.getPropertyValue('--number-of-columns');
    const numberOfCategories = stateStore.categoryList.length;
    document.body.style.setProperty('--number-of-rows', numberOfCategories / numberOfColumns);
  };

  useEffect(() => {
    setRows();
    const disposer = reaction(
      () => [stateStore.screenInfo.width, stateStore.screenInfo.height],
      () => setRows()
    );
    return () => {
      disposer();
    }
  });

  const _mapCategoryList = () => {
      return stateStore.categoryList.map((category, index) => { 
        //Display the Category component, and pass along the showSpecificCategory function as a prop, so that we can call it from the Category component
        return ( <CategoryThumbnail key={category.id.toString()} id={category.id} index={index} name={category.name} stateStore={stateStore} api={api} /> ); }
      );
  }
  
  return (
    <div className="category-list">
      {_mapCategoryList()} 
    </div>
  );
};

Categories.propTypes = {
  stateStore: PropTypes.object.isRequired,
  api: PropTypes.object.isRequired
};

Categories.displayName = 'Categories';

export default observer(Categories);