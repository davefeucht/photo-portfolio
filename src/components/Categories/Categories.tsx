/** ***************
* Categories component displays a list of existing photo categories 
* and makes them clickable to display just a single category.
**************** */

import './Categories.css';

import { observer } from 'mobx-react';
import * as React from 'react';
import { useEffect, useState } from 'react';

import { Category, ScreenInfo, SiteInfo } from '../../utils/types';
import CategoryThumbnail from '../CategoryThumbnail/CategoryThumbnail';
import PaginationNavigation from '../PaginationNavigation/PaginationNavigation';

interface CategoriesProps {
    maxItemsPerPage: number,
    screenInfo: ScreenInfo,
    categoryList: Category[]
}

const Categories: React.FC<CategoriesProps> = ({
    maxItemsPerPage,
    screenInfo,
    categoryList
}) => {
    const [currentPageIndex, setCurrentPageIndex] = useState<number>(1);

    const startIndex = (currentPageIndex - 1) * maxItemsPerPage;
    const endIndex = startIndex + maxItemsPerPage;

    const setRows = () => {
        const numberOfColumns = parseInt(document.body.style.getPropertyValue('--number-of-columns')) || 1;
        const numberOfCategories = maxItemsPerPage;
        document.body.style.setProperty('--number-of-rows', `${numberOfCategories / numberOfColumns}`);
    };

    useEffect(() => {
        setRows();
    }, [screenInfo.width, screenInfo.height]);

    return (
        <div className="category-list">
            {categoryList.slice(startIndex, endIndex)
                .map(category => {
                    return (<CategoryThumbnail key={category.id.toString()} id={category.id} name={category.name} />);
                })}
            <PaginationNavigation totalPages={categoryList.length / maxItemsPerPage} currentPageIndex={currentPageIndex} navigationFunction={(content: number) => { setCurrentPageIndex(content); }} />
        </div>
    );
};

Categories.displayName = 'Categories';

export default observer(Categories);
