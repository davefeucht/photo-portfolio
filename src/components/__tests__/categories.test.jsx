import { render, screen } from '@testing-library/react';
import React from 'react';
import { act } from 'react-dom/test-utils';
import { MemoryRouter, Route, Routes } from 'react-router-dom';

import StateStore from '../../StateStore/store';
import Categories from '../Categories/Categories';

const screenInfo = {
    width: 500,
    height: 500
};

const siteInfo = {
    siteName: 'Through a Pinhole',
    siteUrl: 'throughapinhole.com'
};

const categories = [
    {
        id: 23,
        count: 6,
        description: 'Category of other things',
        link: '',
        name: 'Some other category',
        slug: 'smocg',
        taxonomy: '',
        parent: 12,
        meta: {}
    },
    {
        id: 35,
        count: 3,
        description: 'Category of things',
        link: '',
        name: 'Some category',
        slug: 'smcg',
        taxonomy: '',
        parent: 23,
        meta: {}
    }
];

jest.mock('../../utils/Api', () => ({
    getCategoryImage: () => Promise.resolve('https://throughapinhole.com/wp-content/uploads/2018/07/DSC_1508.jpg')
}));

test('Category list displays', async () => {
    const store = new StateStore();
    store.setCategoryList(categories);
    let container;
    await act(async () => {
        container = render(
            <MemoryRouter>
                <Routes>
                    <Route
                        path="/"
                        element={(
                            <Categories
                                maxItemsPerPage={10}
                                screenInfo={screenInfo}
                                categoryList={categories}
                                siteInfo={siteInfo}
                            />
                        )}
                    />
                </Routes>
            </MemoryRouter>
        ).container;
    });
    expect(container.firstChild).toMatchSnapshot();
    categories.forEach(category => {
        expect(screen.getByText(category.name));
    });
});
