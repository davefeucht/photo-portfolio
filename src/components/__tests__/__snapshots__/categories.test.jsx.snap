// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category list displays 1`] = `
<div
  class="category-list"
>
  <a
    href="/category/23"
  >
    <div
      class="category-thumbnail"
      style="background-image: url(https://throughapinhole.com/wp-content/uploads/2018/07/DSC_1508.jpg);"
    >
      <div
        class="category-title"
      >
        <div
          class="category-title-text"
        >
          Some other category
        </div>
      </div>
    </div>
  </a>
  <a
    href="/category/35"
  >
    <div
      class="category-thumbnail"
      style="background-image: url(https://throughapinhole.com/wp-content/uploads/2018/07/DSC_1508.jpg);"
    >
      <div
        class="category-title"
      >
        <div
          class="category-title-text"
        >
          Some category
        </div>
      </div>
    </div>
  </a>
  <div
    class="pagination-navigation"
  >
    <div
      class="pagination-navigation__content"
    >
      <div
        class="pagination-navigation__item selected"
      >
        1
      </div>
    </div>
  </div>
</div>
`;
